configure.ac

```diff
@@ -221,17 +233,27 @@ AC_ARG_ENABLE([picky-compiler],
 AS_IF([test "${enable_picky_compiler}" != "no"],
       [AC_MSG_NOTICE([Adding ${picky_cflags} to CFLAGS.])
        CFLAGS="${CFLAGS} ${picky_cflags}"
+       CXXFLAGS="${CXXFLAGS} ${picky_cxxflags}"
+       AS_UNSET([picky_cxxflags])
        AS_UNSET([picky_cflags])])
 
 werror_flags="-Werror"
 AC_ARG_ENABLE([werror],
    [AS_HELP_STRING([--enable-werror], [(Developer) Enable setting -Werror.  Off by default, unless building from Git tree.])])
 AS_IF([test -d "${srcdir}/.git" -a -z "${enable_werror}"],
       [AC_MSG_NOTICE([Found .git directory.  Adding ${werror_flags} to CFLAGS.])
-       CFLAGS="${werror_flags} ${CFLAGS}"],
+       CXXFLAGS="${CXXFLAGS} ${werror_flags} "
+       CFLAGS="${CFLAGS} ${werror_flags} "],
       [test "${enable_werror}" = "yes"],
       [AC_MSG_NOTICE([Adding ${werror_flags} to CFLAGS.])
-       CFLAGS="${werror_flags} ${CFLAGS}"])
+       CXXFLAGS="${CXXFLAGS} ${werror_flags} "
+       CFLAGS="${CFLAGS} ${werror_flags} "])
+
+AC_ARG_ENABLE([cpp],
```

Comment:
In autoconf context, cpp is the C pre processor.  This should be `--enable-cxx` or similar.  But I've said a bunch of times, I don't want this half-way situation.  Make the change or don't.  Commit to it.

Reply:
> In autoconf context, cpp is the C pre processor. This should be --enable-cxx or similar.

I've renamed the flag to cxx.

> But I've said a bunch of times, I don't want this half-way situation. Make the change or don't. Commit to it.

`--enable-cxx` flag isn't here because I think it's reasonable to support compiling under either C or C++ longterm. It exclusively exists so that we can run it on every commit and ensure that as someone develops commits targeting C++, they don't need to either fix a bunch of problems in the file before even getting started, nor do they need to go make changes in unrelated files just to get their branch to build again. `--enable-cxx` is disabled by default (regardless of the existence of a .git directory) and exists only for the sake of CI. Unless you specifically enable it, you will exclusively use a C compiler and all files will be built as C. (The exception is the unit tests, which doesn't care either way if you pass `--enable-cxx`, because they've been renamed with a C++ extension and always build with a C++ compiler.)  

I don't understand how to make progress developing C++ changes for the entire project atop a moving master branch that is introducing new breaking changes regularly. The ground is always going to move slightly, but this considerably stops the shaking. I've had this working for two months and have been left holding the door (`AuthorDate: Mon Aug 12 12:37:04 2024 -0700`). 

If we instead had just accepted it at the time, we wouldn't have needed all this cleanup. But that's spilled milk under the bridge -- I've spent the time to maintain the branch and it's here again. It's still just as essential as it was then. It's the prerequisite to actually begin the real work. It's always going to be the prerequisite to actually doing the real work until we get to the point where the C++ build is the only build.

> I don't want this half-way situation

What I think you're getting at is that if the files are able to be renamed to C++, you'd prefer to just do it, and then clean them up later. This is still a "half-way situation". The fact that the code is a certain way right now, and we want it to be a different way in the future, means that there's always going to be a half way situation.

In my mind, doing `git ls-files | xargs rename .c .cc` and having a bunch of C code that's parsing as C++ without actually being C++ is a much worse option. Just because it parses doesn't mean it's valid, and I have a healthy fear of the C++ language lawyers and their ability to motivate compiler teams to do very scary things. 

Take for example this wg21 example: [P0593R5](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0593r5.html#motivating-examples)

> ```c++
> struct X { int a, b; };
> X *make_x() {
>   X *p = (X*)malloc(sizeof(struct X));
>   p->a = 1;
>   p->b = 2;
>   return p;
> }
> ```
> 
> When compiled with a C++ compiler, this code has undefined behavior, because p->a attempts to write to an int subobject of an X object, and this program never created either an X object nor an int subobject.

At the very minimum, every usage of malloc/calloc/etc. in our codebase needs to be replaced with a wrapper that performs `placement new`, or something like `std::bit_cast`, or some sort of smart pointer.

The follow up example is also worth posting:

> Suppose a C++ program is given a sequence of bytes (perhaps from disk or from a network), and it knows those bytes are a valid representation of type T. How can it efficiently obtain a T * that can be legitimately used to access the object?
> 
> Example: (many details omitted for brevity)
> ```c++
> void process(Stream *stream) {
>   unique_ptr<char[]> buffer = stream->read();
>   if (buffer[0] == FOO)
>     process_foo(reinterpret_cast<Foo*>(buffer.get())); // #1
>   else
>     process_bar(reinterpret_cast<Bar*>(buffer.get())); // #2
> }
> ```
> 
> This code leads to undefined behavior today: within Stream::read, no Foo or Bar object is created, and so any attempt to access a Foo object through the Foo* produced by the cast at `#1` would result in undefined behavior.

What's being described here is the core datapath of this codebase -- `fi_cq_read` followed by reinterpretting various fields of the filled `fi_cq_data_entry` as some unrelated type.

Suppose we test a commit that does a big rename and it mostly works. We then pick bits and pieces of C++ that we want, ignore the others, and 6 months pass. All of the work that we never did becomes a ticking time bomb. What are we going to do if we suddenly find ourselves crashing in extremely bizarre ways, not just a month later when we can just revert the commits, but 6 months later when it's too tricky to untangle? I don't want to find myself looking at compiler optimization passes to try to figure out what's wrong, nor do I want to waste time reading up on C++ standards to try to audit the codebase for what we're doing that's unsafe. 

That's the tail risk nightmare I'm trying to mitigate with this.

The C ABI that we're already using everywhere makes it easy for us to keep existing code exactly the way it is, built exactly the same way it was, while interopping bidirectionally between newly added C++ code and our existing C code. When we port any given translation unit to C++, we can just write idiomatic c++ code that avoids these traps in the first place. Consumers of that file see the same C ABI and API they saw previously and need not know that the file changed at all. Changes can be made in isolation and nobody will ever find themselves in a position of pushing a hasty change in file X, because file Y wanted to use std::vector.

Yes, it's unfortunate that some files are going to be in C++ and some are going to be in C and we're going to have to context switch for a month or two. But at least it's just annoying and not actively dangerous.

This PR is just trying to implement a CI job that can support this approach, by making sure that the ignored corners of the codebase doesn't ever drift too far away from being ported. The end goal is definitely to get to the point where this flag is removed and all files are just C++. I just don't want to rush into doing that and find ourselves regretting it.

Reply:
I've said many times that what we're going to do is exactly what you don't want.  Get the code compiling with a C++ compiler, and clean it up iteratively.  Because no one can actually start using C++ until then, and you're holding up everyone trying to make reasonable changes.